### PR TITLE
Make Theme.of non-nullable

### LINKED ISSUE #67163: Consider Theme.of()! and other related topics
With the null safety migration, we've added a lot of `!`s near `of` methods. This issue is to track what we should do about it.

cc @goderbauer @HansMuller @gspencergoog @a14n

### PR DESCRIPTION
**Pull request by goderbauer**
_Note to reviewers: `theme.dart` contains the only interesting change. Everything else is mechanical **removal of 280 `!`s**._

## Description

Make `Theme.of` non-nullable by removing the `shadowThemeOnly` param. That param has been replaced with `InhertiedTheme.capture`, see https://github.com/flutter/flutter/pull/69050.

## Related Issues

Fixes #67163.

## Tests

I added the following tests:

n/a

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] I wrote a design doc: https://flutter.dev/go/template https://docs.google.com/document/d/1LKBL0S1uyhACDiUXILhLKFZH8Cpl_w3mXjHgaE8_--U/edit?usp=sharing
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/inherited_theme.dart [hunk 1]
```diff
@@ -95,27 +99,72 @@ abstract class InheritedTheme extends InheritedWidget {
...
+  ///
+  /// After calling this method, the themes present between `from` and `to` are
+  /// frozen in the returned [CaptureThemes] object. If the themes (or their
+  /// theme data) change in the original subtree, those changes will not be
+  /// applied to the themes captured in the [CaptureThemes] object - unless this
+  /// method is called again to re-capture the updated themes.
+  ///
+  /// To wrap a [Widget] in the captured themes, call [CapturedThemes.wrap].
+  static CapturedThemes capture({ required BuildContext from, required BuildContext? to }) {
+    assert(from != null);
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/inherited_theme.dart:146]: This is necessary with NNBD?

File: packages/flutter/lib/src/widgets/inherited_theme.dart [hunk 2]
```diff
@@ -130,7 +179,23 @@ abstract class InheritedTheme extends InheritedWidget {
       return true;
     });
 
-    return _CaptureAll(themes: themes, child: child);
+    assert(debugDidFindAncestor, 'The provided `to` context must be an ancestor of the `from` context.');
+    return CapturedThemes._(themes);
+  }
+}
+
+/// Stores a list of captured [InheritedTheme] that can be wrapped around a
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/inherited_theme.dart:187]: `[InheritedTheme]`s

File: packages/flutter/lib/src/widgets/inherited_theme.dart [hunk 3]
```diff
@@ -43,19 +41,25 @@ import 'framework.dart';
 /// class MyAppBody extends StatelessWidget {
 ///   @override
 ///   Widget build(BuildContext context) {
+///     final NavigatorState navigator = Navigator.of(context);
+///     // This InheritedTheme.capture() saves references to themes that are
+///     // found above the context provided to this widget's build method
+///     // excluding themes are are found above the navigator. Those themes do
+///     // not have to be captured, because they will already be visible from
+///     // the new route pushed onto said navigator.
+///     // Themes are captures outside of the route's builder because when the
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/inherited_theme.dart:50]: captures => captured

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/inherited_theme.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/inherited_theme.dart:146]: While our customers still run without NNBD we still need them.

**Comment by goderbauer** [comment]
"Google testing" failures are from an unrelated change.

**Pull request merged**